### PR TITLE
DTensor: C++ compute_global_tensor_info

### PR DESCRIPTION
**Pull request by swolchok**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.12.0) (oldest at bottom):
* #163667
* __->__ #162990

compute_global_tensor_info is on the hot path for DTensor.{from,to}_local. More incremental progress toward C++.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @ezyang @msaroufim @dcci @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @Lucaskabela

### CODE REVIEW

File: torch/csrc/autograd/python_variable.cpp [hunk 1]
```diff
@@ -1122,6 +1135,124 @@ static PyObject* DTensor_OpSchema_post_init(PyObject* mod, PyObject* self) {
...
+  py::object mesh_size;
+  for (const auto& placement : placements) {
+    // TODO: C++ify Placement and DeviceMesh somehow; profiling seems
+    // to say that nearly all our remaining time spent is spent
+    // calling back into Python.
+    if (checked_is_true(
+            placement.attr(dtensor_interned_strings.is_shard)().ptr())) {
+      const auto shard_dim =
+          py::cast<int64_t>(placement.attr(dtensor_interned_strings.dim));
+      TORCH_CHECK(
```
Comment (Skylion007) [inline torch/csrc/autograd/python_variable.cpp:1173]: Don't we have a TORCH_CHECK_ASSERT to raise an Assertion Error?

File: torch/csrc/autograd/python_variable.cpp [hunk 2]
```diff
@@ -1122,6 +1135,124 @@ static PyObject* DTensor_OpSchema_post_init(PyObject* mod, PyObject* self) {
...
+    }
+    idx++;
+  }
+  return py::make_tuple(
+             int_array_to_list(tensor_shape), int_array_to_list(tensor_strides))
+      .release()
+      .ptr();
+}
+
+static const char compute_global_tensor_info_doc[] =
```
Comment (Skylion007) [inline torch/csrc/autograd/python_variable.cpp]: Make it constexpr too

File: torch/csrc/autograd/python_variable.cpp [hunk 3]
```diff
@@ -1122,6 +1135,124 @@ static PyObject* DTensor_OpSchema_post_init(PyObject* mod, PyObject* self) {
...
+    "    tensor_shape: A List of int which specifies the size of DTensor which build\n"
+    "        on top of the local tensor.\n"
+    "    tensor_stride: A List of int which specifies the stride of DTensor.\n";
+
+static PyObject* DTensor_compute_global_tensor_info(
+    PyObject* self,
+    PyObject* const* args,
+    Py_ssize_t nargs) {
+  HANDLE_TH_ERRORS
+  TORCH_CHECK(
```
Comment (Skylion007) [inline torch/csrc/autograd/python_variable.cpp]: TORCH_CHECK_VALUE

File: torch/csrc/autograd/python_variable.cpp [hunk 4]
```diff
@@ -1122,6 +1135,124 @@ static PyObject* DTensor_OpSchema_post_init(PyObject* mod, PyObject* self) {
...
+        }
+      }
+    } else if (
+        checked_not(
+            placement.attr(dtensor_interned_strings.is_replicate)().ptr()) &&
+        checked_not(
+            placement.attr(dtensor_interned_strings.is_partial)().ptr())) {
+      const auto placement_type_name =
+          py::str(py::handle(PyType_GetName(Py_TYPE(placement.ptr()))));
+      return PyErr_Format(
```
Comment (Skylion007) [inline torch/csrc/autograd/python_variable.cpp:1209]: You want to return the Error instead of throwing it?

Comment (swolchok) [reply torch/csrc/autograd/python_variable.cpp:1209]: This does raise an exception:
- PyErr_Format flags an error
- PyErr_Format always returns nullptr, which is what you return while raising an exception
- the previous PR did add a test for this

File: torch/csrc/autograd/python_variable.cpp (hunk 2 above)
Comment (swolchok) [reply torch/csrc/autograd/python_variable.cpp]: OK, that seems perfectly reasonable. note that it won't do anything in particular. constexpr is a "may", not a "must", and the `static const` part will already permit this to be initialized at compile time.

File: torch/csrc/autograd/python_variable.cpp (hunk 1 above)
Comment (swolchok) [reply torch/csrc/autograd/python_variable.cpp:1173]: not that I found. TORCH_INTERNAL_ASSERT exists but doesn't seem to do anything differently w.r.t. the exception it raises.

**Comment by swolchok** [comment]
> will try TORCH_LIBRARY

Hmm, making this a custom op seems bad, at least right now. The function signature includes DeviceMesh (not currently IValue-able) and a list of Placements (also not currently IValue-able). I guess maybe I'm supposed to 1) add the Placement stuff from https://github.com/pytorch/pytorch/pull/163030 to IValue 2) have a C++ DeviceMesh representation 3) add that DeviceMesh representation to IValue 4) teach the custom operator / native function schema stuff about all that and then 5) sequence this PR after all those things?

On the other hand, torch.compiler.allow_in_graph might make this easy. I'll try that first.

**Comment by swolchok** [comment]
> On the other hand, torch.compiler.allow_in_graph might make this easy. I'll try that first.

Simply adding
```
        import torch._C
        torch.compiler.allow_in_graph(torch._C._DTensor_compute_global_tensor_info)
```

to the start of test_dtensor_dynamic does not cause it to pass. @ezyang does it sound right that C++ifying utility functions that happen to work with Tensors has to be done through custom operators, which necessitates making all their argument types (i.e., DeviceMesh, Placement) go through IValue?

**Comment by ezyang** [comment]
They look ignorable. Is there a reason you put periodic label on it?

File: test/inductor/test_distributed_patterns.py [hunk 5]
```diff
@@ -3,6 +3,7 @@
 import functools
 
 import torch
+import torch.distributed.fsdp
```
Comment (ezyang) [inline test/inductor/test_distributed_patterns.py]: weird

File: torch/csrc/utils/python_compat.h [hunk 6]
```diff
@@ -9,11 +9,11 @@ extern "C" {
...
-#define IS_PYTHON_3_11_PLUS PY_VERSION_HEX >= 0x030B00C1
-#define IS_PYTHON_3_12_PLUS PY_VERSION_HEX >= 0x030C0000
-#define IS_PYTHON_3_13_PLUS PY_VERSION_HEX >= 0x030D0000
-#define IS_PYTHON_3_14_PLUS PY_VERSION_HEX >= 0x030E0000
-#define IS_PYTHON_3_15_PLUS PY_VERSION_HEX >= 0x030F0000
+#define IS_PYTHON_3_11_PLUS (PY_VERSION_HEX >= 0x030B00C1)
+#define IS_PYTHON_3_12_PLUS (PY_VERSION_HEX >= 0x030C0000)
+#define IS_PYTHON_3_13_PLUS (PY_VERSION_HEX >= 0x030D0000)
+#define IS_PYTHON_3_14_PLUS (PY_VERSION_HEX >= 0x030E0000)
+#define IS_PYTHON_3_15_PLUS (PY_VERSION_HEX >= 0x030F0000)
```
Comment (ezyang) [inline torch/csrc/utils/python_compat.h:16]: 👍

File: torch/csrc/utils/pycfunction_helpers.h [hunk 7]
```diff
@@ -11,3 +12,15 @@ inline PyCFunction castPyCFunctionWithKeywords(PyCFunctionWithKeywords func) {
   C10_DIAGNOSTIC_POP()
   C10_DIAGNOSTIC_POP()
 }
+
+#if !IS_PYTHON_3_13_PLUS
+using PyCFunctionFast = _PyCFunctionFast;
```
Comment (ezyang) [inline torch/csrc/utils/pycfunction_helpers.h:17]: cc @williamwen42

File: torch/csrc/autograd/python_variable.cpp [hunk 8]
```diff
@@ -1130,6 +1140,130 @@ static PyObject* DTensor_OpSchema_post_init(PyObject* mod, PyObject* self) {
...
+    "        Object which describes the mesh topology\n"
+    "        of devices for the DTensor.\n"
+    "    placements (Sequence[:class:`Placement`]]):\n"
+    "        The attribute of the DTensor that describes its layout\n"
+    "        on the mesh topology.\n"
+    "\n"
+    "Return:\n"
+    "    tensor_shape: A List of int which specifies the size of DTensor which build\n"
+    "        on top of the local tensor.\n"
+    "    tensor_stride: A List of int which specifies the stride of DTensor.\n";
```
Comment (ezyang) [inline torch/csrc/autograd/python_variable.cpp:1247]: How do you feel about unit test coverage on this?

File: torch/csrc/autograd/python_variable.cpp [hunk 9]
```diff
@@ -1130,6 +1140,130 @@ static PyObject* DTensor_OpSchema_post_init(PyObject* mod, PyObject* self) {
...
+    Py_ssize_t nargs) {
+  HANDLE_TH_ERRORS
+  TORCH_CHECK_VALUE(
+      nargs == 3,
+      "compute_global_tensor_info expects 3 arguments, got ",
+      nargs);
+  TORCH_CHECK_TYPE(
+      THPVariable_Check(args[0]),
+      "compute_global_tensor_info 1st argument must be Tensor!");
+  const auto& tensor = THPVariable_Unpack(args[0]);
```
Comment (ezyang) [inline torch/csrc/autograd/python_variable.cpp:1261]: ah the joy of const reference lifetime extension

File: torch/csrc/autograd/python_variable.cpp [hunk 10]
```diff
@@ -1130,6 +1140,130 @@ static PyObject* DTensor_OpSchema_post_init(PyObject* mod, PyObject* self) {
...
+  // NOTE: if this is a py::handle then this code stops working;
+  // apparently we can't rely on the bound method to stick around.
+  py::object mesh_size;
+  for (const auto& placement : placements) {
+    // TODO: C++ify DeviceMesh somehow; profiling seems
+    // to say that nearly all our remaining time spent is spent
+    // calling back into Python.
+    const auto& cpp_placement = placement.cast<const distributed::Placement&>();
+    if (const auto* cpp_shard =
+            dynamic_cast<const distributed::Shard*>(&cpp_placement)) {
```
Comment (ezyang) [inline torch/csrc/autograd/python_variable.cpp:1171]: I love the smell of RTTI for breakfast (do we really not have a type tag on these yet?)

File: torch/csrc/autograd/python_variable.cpp [hunk 11]
```diff
@@ -1130,6 +1140,130 @@ static PyObject* DTensor_OpSchema_post_init(PyObject* mod, PyObject* self) {
...
+      }
+      const auto mesh_dim_size = py::cast<int64_t>(mesh_size(idx));
+      tensor_shape[shard_dim] *= mesh_dim_size;
+      for (const auto i : c10::irange(tensor_strides.size())) {
+        if (static_cast<int64_t>(i) != shard_dim &&
+            tensor_strides[i] >= tensor_strides[shard_dim]) {
+          tensor_strides[i] *= mesh_dim_size;
+        }
+      }
+    } else if (!cpp_placement.is_replicate() && !cpp_placement.is_partial()) {
```
Comment (ezyang) [inline torch/csrc/autograd/python_variable.cpp:1200]: FWIW, I kind of suspect the logic is wrong for testing this way, and this should just be an else, but that'd be semantics changing so best to do it separately.

File: torch/csrc/autograd/python_variable.cpp [hunk 12]
```diff
@@ -1130,6 +1140,130 @@ static PyObject* DTensor_OpSchema_post_init(PyObject* mod, PyObject* self) {
...
+             symint_array_to_list(tensor_shape),
+             symint_array_to_list(tensor_strides))
+      .release()
+      .ptr();
+}
+
+// NOLINTNEXTLINE(modernize-avoid-c-arrays,cppcoreguidelines-avoid-c-arrays)
+static constexpr const char compute_global_tensor_info_doc[] =
+    "Compute the global size and stride of a DTensor from the given local tensor.\n"
+    "The local size is multiplied by `world_size` per Sharding dim.\n"
```
Comment (ezyang) [inline torch/csrc/autograd/python_variable.cpp:1226]: thanks for fixing the typo lol

File: torch/csrc/autograd/python_variable.cpp [hunk 13]
```diff
@@ -1130,6 +1140,130 @@ static PyObject* DTensor_OpSchema_post_init(PyObject* mod, PyObject* self) {
...
+  Py_ssize_t idx = 0;
+  c10::SymDimVector tensor_shape(
+      tensor.sym_sizes().begin(), tensor.sym_sizes().end());
+  c10::SymDimVector tensor_strides(
+      tensor.sym_strides().begin(), tensor.sym_strides().end());
+  // NOTE: if this is a py::handle then this code stops working;
+  // apparently we can't rely on the bound method to stick around.
+  py::object mesh_size;
+  for (const auto& placement : placements) {
+    // TODO: C++ify DeviceMesh somehow; profiling seems
```
Comment (ezyang) [inline torch/csrc/autograd/python_variable.cpp:1166]: cc @fduwjj @lw 

In this case, you don't actually need the full mesh: you only need the size which is basically an IntArrayRef.

File: torch/csrc/autograd/python_variable.cpp [hunk 14]
```diff
@@ -1130,6 +1140,130 @@ static PyObject* DTensor_OpSchema_post_init(PyObject* mod, PyObject* self) {
...
+    // to say that nearly all our remaining time spent is spent
+    // calling back into Python.
+    const auto& cpp_placement = placement.cast<const distributed::Placement&>();
+    if (const auto* cpp_shard =
+            dynamic_cast<const distributed::Shard*>(&cpp_placement)) {
+      const auto shard_dim = cpp_shard->dim;
+      TORCH_CHECK(
+          shard_dim >= 0,
+          "Shard placements should have negative dims normalized in the user-facing APIs: ",
+          py::cast<std::string>(py::str(placement)));
```
Comment (ezyang) [inline torch/csrc/autograd/python_variable.cpp:1176]: this is annoyingly wordy, printing placement directly ought to work

Comment (ezyang) [reply torch/csrc/autograd/python_variable.cpp:1176]: I guess the problem is it's a Python object so you can't cuz it would implicitly have to take GIL :(

File: torch/csrc/autograd/python_variable.cpp [hunk 15]
```diff
@@ -1130,6 +1140,130 @@ static PyObject* DTensor_OpSchema_post_init(PyObject* mod, PyObject* self) {
...
+          tensor_ndim,
+          " for placement number ",
+          idx);
+
+      if (!mesh_size) {
+        mesh_size = mesh.attr(dtensor_interned_strings.size);
+      }
+      const auto mesh_dim_size = py::cast<int64_t>(mesh_size(idx));
+      tensor_shape[shard_dim] *= mesh_dim_size;
+      for (const auto i : c10::irange(tensor_strides.size())) {
```
Comment (ezyang) [inline torch/csrc/autograd/python_variable.cpp:1194]: Can we preserve the comments from the old code? thx

**Review by ezyang — APPROVED** [review]

File: torch/csrc/autograd/python_variable.cpp (hunk 9 above)
Comment (swolchok) [reply torch/csrc/autograd/python_variable.cpp:1261]: it's not lifetime extension; unpack returns a reference

File: test/inductor/test_distributed_patterns.py (hunk 5 above)
Comment (swolchok) [reply test/inductor/test_distributed_patterns.py]: this test uses ops that that module defines

File: torch/csrc/autograd/python_variable.cpp (hunk 10 above)
Comment (swolchok) [reply torch/csrc/autograd/python_variable.cpp:1171]: Hmm. It's been long enough that I'm not 100% sure what I was thinking, but.I now note that if this was just an  `is_shard()` call and someone has written a Python class that is a subclass of Placement but not the C++ shard class and yet is_shard returns true for it, sadness would ensue

File: torch/csrc/autograd/python_variable.cpp (hunk 8 above)
Comment (swolchok) [reply torch/csrc/autograd/python_variable.cpp:1247]: there appear to be tests in test/distributed/tensor/test_utils.py

**Comment by swolchok** [comment]
failures this time around:
-  [linux-jammy-cuda12.8-py3.10-gcc11 / test (default, 4, 5, linux.g6.4xlarge.experimental.nvidia.gpu)](https://github.com/pytorch/pytorch/actions/runs/18918352725/job/54010750231?pr=162990#logs) looks like it was also segfaulting on main around the time of the base commit: https://github.com/pytorch/pytorch/actions/runs/18862532009/job/53825651259
- the executorch job is marked unstable and seems quite unlikely to be related


as best I can tell we're good to merge with -i.

**Comment by jithunnair-amd** [comment]
@swolchok You might want to use `merge -f` if the ROCm distributed jobs are the only ones pending, since the ROCm distributed jobs running in the latest round are looking for the label `linux.rocm.gpu.mi250.4`, and will probably be queued for a long time.
